pkg/apis/v1beta1/aksnodeclass.go

```diff
@@ -0,0 +1,188 @@
+	// +optional
+	ImageGCLowThresholdPercent *int32 `json:"imageGCLowThresholdPercent,omitempty"`
+	// topologyManagerPolicy is the name of the topology manager policy to use.
+	// Valid values include:
+	//
+	// - `restricted`: kubelet only allows pods with optimal NUMA node alignment for requested resources;
+	// - `best-effort`: kubelet will favor pods with NUMA alignment of CPU and device resources;
+	// - `none`: kubelet has no knowledge of NUMA alignment of a pod's CPU and device resources.
+	// - `single-numa-node`: kubelet only allows pods with a single NUMA alignment
+	//   of CPU and device resources.
+	//
+	// +kubebuilder:validation:Enum:={restricted,best-effort,none,single-numa-node}
+	// +kubebuilder:default="none"
+	// +optional
+	TopologyManagerPolicy string `json:"topologyManagerPolicy,omitempty"`
+	// A comma separated whitelist of unsafe sysctls or sysctl patterns (ending in `*`).
+	// Unsafe sysctl groups are `kernel.shm*`, `kernel.msg*`, `kernel.sem`, `fs.mqueue.*`,
+	// and `net.*`. For example: "`kernel.msg*,net.ipv4.route.min_pmtu`"
+	// Default: []
+	// TODO: validation
+	// +optional
+	AllowedUnsafeSysctls []string `json:"allowedUnsafeSysctls,omitempty"`
+	// containerLogMaxSize is a quantity defining the maximum size of the container log
+	// file before it is rotated. For example: "5Mi" or "256Ki".
+	// Default: "10Mi"
+	// AKS CustomKubeletConfig has containerLogMaxSizeMB (with units), defaults to 50
+	// +kubebuilder:validation:Pattern=`^\d+(E|P|T|G|M|K|Ei|Pi|Ti|Gi|Mi|Ki)$`
+	// +kubebuilder:default="50Mi"
+	// +optional
+	ContainerLogMaxSize string `json:"containerLogMaxSize,omitempty"`
+	// containerLogMaxFiles specifies the maximum number of container log files that can be present for a container.
+	// Default: 5
+	// +kubebuilder:validation:Minimum:=2
+	// +kubebuilder:default=5
+	// +optional
+	ContainerLogMaxFiles *int32 `json:"containerLogMaxFiles,omitempty"`
+	// podPidsLimit is the maximum number of PIDs in any pod.
+	// AKS CustomKubeletConfig uses PodMaxPids, int32 (!)
+	// Default: -1
+	// +optional
+	PodPidsLimit *int64 `json:"podPidsLimit,omitempty"`
+}
+
+// AKSNodeClass is the Schema for the AKSNodeClass API
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=aksnodeclasses,scope=Cluster,categories={karpenter,nap},shortName={aksnc,aksncs}
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+type AKSNodeClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   AKSNodeClassSpec   `json:"spec,omitempty"`
+	Status AKSNodeClassStatus `json:"status,omitempty"`
+}
+
+// We need to bump the AKSNodeClassHashVersion when we make an update to the AKSNodeClass CRD under these conditions:
+// 1. A field changes its default value for an existing field that is already hashed
+// 2. A field is added to the hash calculation with an already-set value
+// 3. A field is removed from the hash calculations
+const AKSNodeClassHashVersion = "v2"
```

Comment:
This feels like something that actually shouldn't be in the API package at all, because it's about how we internally version our hashing mechanism - which doesn't/shouldn't vary per-api version?

Reply:
The hashing algorithm version is very closely tied to which fields are included or not, and is likely to change between API versions (as new fields are introduces) - except when we create an exact replica ...

Reply:
but the operator only acts on the latest storage version, which always has "all" the fields. So even if we add fields and they're newly included in the new API version, they'll still be "under consideration" even for the older API versions, because at the reconciler level we aren't considering the shape the user interacted with, we're considering the canonical shape.

Reply:
which is my long-winded way of saying: "The version of the hash algorithm is tied to the backing controller, not the API version of the NodeClass. Even if we had 27 versions of NodeClass there would only be a single HashVersion.

Reply:
I filed https://github.com/Azure/karpenter-poc/issues/1234 - we can punt on this for this PR and somebody can fix it later. I assigned it to me for now but that's mostly just so I don't lose track of it.

Reply:
thanks @matthchr